naming/program/src/state.rs

```diff
@@ -0,0 +1,47 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{msg, program_error::ProgramError, pubkey::Pubkey};
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum NameStatus {
+    Deleted,
+    Active,
+}
+
+pub type ValueType = Pubkey;
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone, Copy)]
+pub enum IssuerKind {
+    Metadata,       // Metadata authority has sole control
+    Issuer(Pubkey), // Issuing authority has sole control
+    Both(Pubkey),   // Issuing and metadata authority must sign
+    Either(Pubkey), // Issuing or metadata authority must sign
+    Open,           // First come, first served. Terminal state
+    Frozen,         // No further issuance allowed.  Terminal state
+}
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+pub enum OwnerKind {
+    Account(Pubkey),
+    Name(String), // Usefull for easy ownership transfer of a whole sub-tree of domain names
+    Frozen,       // Metadata is locked, terminal state
+}
+
+#[derive(Debug, PartialEq, BorshDeserialize, BorshSerialize, Clone)]
+pub struct NameRegistryState {
+    pub status: NameStatus,
+    // This is the complete name, including all parents
+    pub name: String,
+    pub metadata_authority: OwnerKind,
+    pub issuing_authority: IssuerKind,
```

Comment:
It would be nice to combine these two fields, and get to a single Owner Kind - "Pubkey".  Frozen then becomes `Pubkey::default()`.